napari/layers/shapes/_shapes_mouse_bindings.py

```diff
@@ -0,0 +1,272 @@
+    """Add an ellipse."""
+    size = layer._vertex_size * layer.scale_factor / 4
+    corner = np.array(layer.displayed_coordinates)
+    data = np.array(
+        [corner, corner + [size, 0], corner + size, corner + [0, size]]
+    )
+    yield from _add_line_rectangle_ellipse(
+        layer, event, data=data, shape_type='ellipse'
+    )
+
+
+def add_rectangle(layer, event):
+    """Add an rectangle."""
+    size = layer._vertex_size * layer.scale_factor / 4
+    corner = np.array(layer.displayed_coordinates)
+    data = np.array(
+        [corner, corner + [size, 0], corner + size, corner + [0, size]]
+    )
+    yield from _add_line_rectangle_ellipse(
+        layer, event, data=data, shape_type='rectangle'
+    )
+
+
+def _add_line_rectangle_ellipse(layer, event, data, shape_type):
+    """Helper function for adding a a line, rectangle or ellipse."""
+
+    # on press
+    # Start drawing rectangle / ellipse / line
+    data_full = layer.expand_shape(data)
+    layer.add(data_full, shape_type=shape_type)
+    layer.selected_data = {layer.nshapes - 1}
+    layer._value = (layer.nshapes - 1, 4)
+    layer._moving_value = copy(layer._value)
+    layer.refresh()
+    yield
+
+    # on move
+    while event.type == 'mouse_move':
+        # Drag any selected shapes
+        layer._move(layer.displayed_coordinates)
+        yield
+
+    # on release
+    layer._finish_drawing()
+
+
+def add_path_polygon(layer, event):
+    """Add a path or polygon."""
+    coord = layer.displayed_coordinates
+
+    # on press
+    if layer._is_creating is False:
+        # Start drawing a path
+        data = np.array([coord, coord])
+        data_full = layer.expand_shape(data)
+        layer.add(data_full, shape_type='path')
+        layer.selected_data = {layer.nshapes - 1}
+        layer._value = (layer.nshapes - 1, 1)
+        layer._moving_value = copy(layer._value)
+        layer._is_creating = True
+        layer._set_highlight()
+    else:
+        # Add to an existing path or polygon
+        index = layer._moving_value[0]
+        if layer._mode == Mode.ADD_POLYGON:
+            new_type = Polygon
+        else:
+            new_type = None
+        vertices = layer._data_view.displayed_vertices[
+            layer._data_view.displayed_index == index
+        ]
+        vertices = np.concatenate((vertices, [coord]), axis=0)
+        # Change the selected vertex
+        layer._value = (layer._value[0], layer._value[1] + 1)
+        layer._moving_value = copy(layer._value)
+        data_full = layer.expand_shape(vertices)
+        layer._data_view.edit(index, data_full, new_type=new_type)
+        layer._selected_box = layer.interaction_box(layer.selected_data)
+
+
+def add_path_polygon_creating(layer, event):
+    """While a path or polygon move next vertex to be added."""
+    if layer._is_creating:
+        layer._move(layer.displayed_coordinates)
+
+
+def vertex_insert(layer, event):
+    """Insert a vertex into a selected shape."""
```

Comment:
I've cleaned up some code and added a bunch of comments. hopefully this makes it more clear. i'm not sure splitting would help as much

Reply:
I'm going to merge, but just to reiterate a rule of thumb I love: functions should fit in a screen, ie 25-50 lines max, so that you can "at a glance" tell what it's doing. Humans are not good at holding context over 100 lines of code, so by the time I'm looking at the end of the function, I have forgotten what was happening at the start, and I just have to pray that there are not any non-local effects at play.

(Of course, this is not helped if you call a function `_super_complex_helper(var1, var2, ..., var100)` in the middle. But typically you can break it down into *sensible* chunks that can be read easily, e.g. `validate_event(event, attributes_required)` or similar. Then I can instantly guess at what it's supposed to do, trust that it's doing it's job because I will review that bit of self-contained code by itself later, and move on to the rest of the function.)

Reply:
yes for sure, thanks for merging - definitely something to continue to improve